pkg/replica/replica.go

```diff
@@ -1097,6 +1099,53 @@ func (r *Replica) ReadAt(buf []byte, offset int64) (int, error) {
 	return c, err
 }
 
+func (r *Replica) UnmapAt(length uint32, offset int64) (n int, err error) {
+	defer func() {
+		if err != nil {
+			logrus.Errorf("Replica with dir %v failed to do unmap with offset %v and length %v: %v", r.dir, offset, length, err)
+		}
+	}()
+	if r.readOnly {
+		return 0, fmt.Errorf("can not unmap on read-only replica")
+	}
+
+	r.Lock()
+	// Figure out the disk chain that can be unmapped.
+	// The chain starts from the volume head,
+	// and is followed by continuous removed disks.
+	// For list `unmappableDisks`, the first entry is the volume head,
+	// the second one is the parent of the first entry...
+	unmappableDisks := []string{r.diskPath(r.activeDiskData[len(r.activeDiskData)-1].Name)}
+	for idx := len(r.activeDiskData) - 2; idx > 0; idx-- {
+		disk := r.activeDiskData[idx]
+		if len(r.diskChildrenMap[disk.Name]) > 1 {
+			break
+		}
+		if r.unmapMarkDiskChainRemoved {
+			if err := r.markDiskAsRemoved(disk.Name); err != nil {
+				return 0, errors.Wrapf(err, "failed to mark snapshot disk %v as removed before unmap", disk.Name)
+			}
+		}
+		if disk.Removed {
+			unmappableDisks = append(unmappableDisks, r.diskPath(disk.Name))
+		}
+	}
+	r.info.Dirty = true
+	unmappedSize, err := r.volume.UnmapAt(unmappableDisks, length, offset)
+	r.Unlock()
```

Comment:
Should move `r.Unlock()` just next to `r.Lock()` to avoid some error return before `r.Unlock()`

If you don't want to unlock the lock until the end of this method, you can use an anonymous closure to run the above unmap operations and do the lock/unlock inside.

Reply:
Modified. Using an anonymous closure instead.

Reply:
:point_up: this is good